sklearn/model_selection/_classification_threshold.py

```diff
@@ -0,0 +1,924 @@
+            else:
+                # single split cross-validation
+                train_idx, _ = next(cv.split(X, y, **routed_params.splitter.split))
+                X_train = _safe_indexing(X, train_idx)
+                y_train = _safe_indexing(y, train_idx)
+                fit_params_train = _check_method_params(
+                    X, routed_params.estimator.fit, indices=train_idx
+                )
+
+            self.estimator_.fit(X_train, y_train, **fit_params_train)
+
+        if hasattr(self.estimator_, "n_features_in_"):
+            self.n_features_in_ = self.estimator_.n_features_in_
+        if hasattr(self.estimator_, "feature_names_in_"):
+            self.feature_names_in_ = self.estimator_.feature_names_in_
+
+        if self.strategy == "constant":
+            # early exit when we don't need to find the optimal threshold
+            self.best_threshold_ = self.constant_threshold
+            self.best_score_, self.constrained_score_ = None, None
+            if self.store_cv_results:
+                self.cv_results_ = None
+            return self
+
+        cv_thresholds, cv_scores = zip(
+            *Parallel(n_jobs=self.n_jobs)(
+                delayed(_fit_and_score_over_thresholds)(
+                    classifier,
+                    X,
+                    y,
+                    fit_params=routed_params.estimator.fit,
+                    train_idx=train_idx,
+                    val_idx=val_idx,
+                    curve_scorer=self._curve_scorer,
+                    score_method=self.objective_metric,
+                    score_params=routed_params.scorer.score,
+                )
+                for train_idx, val_idx in splits
+            )
+        )
+
+        if any(np.isclose(th[0], th[-1]) for th in cv_thresholds):
+            raise ValueError(
+                "The provided estimator makes constant predictions. Therefore, it is "
+                "impossible to optimize the decision threshold."
+            )
+
+        # find the global min and max thresholds across all folds
+        min_threshold = min(
+            split_thresholds.min() for split_thresholds in cv_thresholds
+        )
+        max_threshold = max(
+            split_thresholds.max() for split_thresholds in cv_thresholds
+        )
+        if isinstance(self.n_thresholds, Integral):
+            decision_thresholds = np.linspace(
+                min_threshold, max_threshold, num=self.n_thresholds
+            )
+        else:
+            decision_thresholds = np.asarray(self.n_thresholds)
+
+        def _mean_interpolated_score(target_thresholds, cv_thresholds, cv_scores):
+            return np.mean(
+                [
+                    np.interp(target_thresholds, split_thresholds, split_score)
+                    for split_thresholds, split_score in zip(cv_thresholds, cv_scores)
+                ],
+                axis=0,
+            )
+
```

Comment:
Why is it not defined outside of the class ? with a docstring :)

Reply:
I believe that if we always use the same thresholds as discussed in a previous comment we don't need that

Reply:
Indeed, we don't need to but I'm not sure of the strategy to define the upper and lower bounds for the `decision_function` case.

Reply:
For the moment I extracted the function and added a docstring.